addons/product/models/product_template.py

```diff
@@ -448,34 +447,49 @@ def create_variant_ids(self):
-            for value_ids in variant_matrix:
-                if set(value_ids.ids) not in existing_variants and not any(value_id.attribute_id.create_variant == 'dynamic' for value_id in value_ids):
-                    variants_to_create.append({
-                        'product_tmpl_id': tmpl_id.id,
-                        'attribute_value_ids': [(6, 0, value_ids.ids)]
-                    })
-
-            if len(variants_to_create) > 1000:
-                raise UserError(_("""
-                The number of variants to generate is too high.
-                You should either not generate variants for each combination or generate them on demand from the sales order.
-                To do so, open the form view of attributes and change the mode of *Create Variants*."""))
-
-            # check product
+            # Determine which product variants need to be created based on the attribute
+            # configuration. If any attribute is set to generate variants dynamically, skip the
+            # process.
+            # Technical note: if there is no attribute, a variant is still created because
+            # 'not any([])' and 'set([]) not in set([])' are True.
+            if not any(create_variant == 'dynamic' for create_variant in tmpl_id.mapped('attribute_line_ids.attribute_id.create_variant')):
+                # Iterator containing all possible attribute values combination
+                # The iterator is used to avoid MemoryError in case of a huge number of combination.
+                all_variants = itertools.product(*(
+                    line.value_ids.ids for line in tmpl_id.attribute_line_ids if line.value_ids[:1].attribute_id.create_variant != 'no_variant'
+                ))
+                # Set containing existing attribute values combination
+                existing_variants = {
+                    frozenset(variant.attribute_value_ids.filtered(lambda r: r.attribute_id.create_variant != 'no_variant').ids)
+                    for variant in tmpl_id.product_variant_ids
+                }
+                # For each possible variant, create if it doesn't exist yet.
+                for value_ids in all_variants:
+                    value_ids = set(value_ids)
+                    if value_ids not in existing_variants:
+                        variants_to_create.append({
+                            'product_tmpl_id': tmpl_id.id,
+                            'attribute_value_ids': [(6, 0, list(value_ids))],
+                        })
+                        if len(variants_to_create) > 1000:
+                            raise UserError(_("""
+                            The number of variants to generate is too high.
+                            You should either not generate variants for each combination or generate them on demand from the sales order.
+                            To do so, open the form view of attributes and change the mode of *Create Variants*."""))
```

Comment:
According to @mart-e translations are not dedented before storage/matching so this is going to change/break the translation.

Reply:
Indeed, see how bad it looks in the po files ;-)
https://github.com/odoo/odoo/blob/dcc077afc523d194589dc007cfd9bc368c9067ca/addons/product/i18n/fr.po#L34

Reply:
OK well, I guess we can still afford to break this one.

Reply:
but then, the previous entry should be removed from the .pot too